core/src/main/scala/org/http4s/parser/Rfc3986Parser.scala

```diff
@@ -6,21 +6,23 @@ import java.net.URLDecoder
 import java.nio.charset.Charset
 import org.http4s.{Query => Q}
 import org.http4s.internal.parboiled2._
+import org.http4s.internal.parboiled2.CharPredicate.{Alpha, Digit, HexDigit}
 import org.http4s.internal.parboiled2.support.HNil
 import org.http4s.syntax.string._
 
-private[parser] trait Rfc3986Parser extends IpParser with StringBuilding {
-  this: Parser =>
-
+private[parser] trait Rfc3986Parser
+    extends Parser
+    with Scheme.Parser
+    with IpParser
+    with StringBuilding {
   // scalastyle:off public.methods.have.type
-  import CharPredicate.{Alpha, Digit, HexDigit}
 
   def charset: Charset
 
   def Uri: Rule1[org.http4s.Uri] = rule { (AbsoluteUri | RelativeRef) ~ EOI }
 
   def AbsoluteUri = rule {
-    Scheme ~ ":" ~ HierPart ~ optional("?" ~ Query) ~ optional("#" ~ Fragment) ~> {
+    scheme ~ ":" ~ HierPart ~ optional("?" ~ Query) ~ optional("#" ~ Fragment) ~> {
```

Comment:
To build these more complex parsers, we're going to have to mix in the rules that know how to construct the types.  I think this is okay.